language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -84,8 +85,36 @@ data BuiltinSized
 data TypedBuiltinSized a where
     TypedBuiltinSizedInt  :: TypedBuiltinSized Integer
     TypedBuiltinSizedBS   :: TypedBuiltinSized BSL.ByteString
-    -- We may actually want to ignore sizes on the Haskell side.
-    TypedBuiltinSizedSize :: TypedBuiltinSized Size
+    -- See Note [Semantics of sizes].
+    TypedBuiltinSizedSize :: TypedBuiltinSized ()
+
+{- Note [Semantics of sizes]
+We convert each PLC's @size s@ into Haskell's '()'. I.e. sizes are completely ignored in
+the semantics of various built-ins. Hence the Haskell's type signature of PLC's 'resizeInteger' is
+
+    () -> Integer -> Integer
+
+while its PLC signature is
+
+    forall s0 s1. size s1 -> integer s0 -> integer s1
+
+This does not mean that we do not perform all the checks prescribed by the specification.
+Merely that we can't compute using sizes on the Haskell side, e.g. we cannot assign a semantics to
+
+    maxIntegerOfGivenSize : forall s. size s -> integer s
+
+There are two reasons for that:
+
+1. clarity. We want to be clear that sizes get passed in another way (see the "Apply" module)
+2. the semantics of @integer s@ in 'Integer'. While this may look ok, we actually lose size information,
```

Comment:
I don't understand what's going on here. How does `() -> Integer -> Size` allow us to compute `sizeOfInteger`?

Reply:
Oh, no, I didn't forget. Anyway, I'll amend it a bit.

Reply:
Done.